Neutron-damaged tungsten validation case from Dark et. al (#42)
festim-dev/V-V-report: report/validation/thermodesorption_spectra/dark/dark_tds.md

Reviewer: There is too much code duplication here. You only need to run each dpa case once, but make an additional plot for one of them. 

You need several functions: 
`run_tds(densities)` : runs a festim model for a given set of **damage-induced** trap densities and returns the `DerivedQuantities` object
`plot_simulated_tds(derived_quantities, trap_contributions: bool, **kwargs)`: plots a simulated TDS from a `DerivedQuantities` object, `kwargs` are the keyword arguments for `plt.plot(x, y, **kwargs)`. If `trap_contributions` is True then also plot the grey areas for each trap.
`plot_experimental_tds(dpa, **kwargs)` : plots the experimental tds based on a dpa level
`scatter_experimental_tds(dpa, **kwargs)`: same as `plot_experimental_tds` but for a scatter plot.

Once you have all these functions, the implementation is simply:

```python
dpa_to_densities = {
    0.001: [1e+24, 2.5e+24, 1e+24, 1e+24, 2e+23], 
    0.005: [3.5e+24, 5e+24, 2.5e+24, 1.9e+24, 1.6e+24], 
    0.023: [2.2e+25, 1.5e+25, 6.5e+24, 2.1e+25, 6e+24], 
    0.1: [4.8e+25, 3.8e+25, 2.6e+25, 3.6e+25, 1.1e+25], 
    0.23: [5.4e+25, 4.4e+25, 3.6e+25, 3.9e+25, 1.4e+25], 
    0.5: [5.5e+25, 4.6e+25, 4e+25, 4.5e+25, 1.7e+25], 
    2.5: [5.8e+25, 6.5e+25, 4.5e+25, 5.5e+25, 2e+25], # re-fit
}

for dpa, densities in dpa_to_densities.items():
    derived_quantities = run_tds(densities)
    plt.figure(1)
    plot_experimental_tds(dpa)
    plot_simulated_tds(derived_quantities)
    if dpa == 0.02:
        plt.figure(2)
        plot_simulated_tds(derived_quantities, trap_contribution=True)
        scatter_experimental_tds(dpa)
```

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.2
kernelspec:
  display_name: vv-festim-report-env
  language: python
  name: python3
---

> # Deuterium retention in neutron-damaged tungsten

Reviewer: ```suggestion
# Deuterium retention in self-damaged tungsten
```

```{tags} 1D, TDS, trapping, transient
```

+++

> This validation case is a thermo-desorption spectrum measurement perfomed by {cite}`dark_modelling_2024`.

Reviewer: ```suggestion
This validation case is a thermo-desorption spectrum measurement on damaged tungsten. The complete description is available in {cite}`dark_modelling_2024`.
```

> Several 0.8 mm thick samples of tungsten were damaged via annealing before being used to perform a TDS measurement.

Reviewer: ```suggestion
Several 0.8 mm thick samples of tungsten were self-damaged and annealing before being used to perform a TDS measurement.
```

> An ion beam with an incident flux of $5.79 \times 10^{19} \ \mathrm{D \ m^{-2} \ s^{-1}}$ was turned on for $72\mathrm{h}$ with an implantation temperature of $370\mathrm{K}$. The sample then rested for $12\mathrm{h}$ at $295\mathrm{K}$ before beggining the TDS measurement at $300\mathrm{K}$ with a temperature ramp of $0.05\mathrm{K}/s$.

Reviewer: ```suggestion
An ion beam with an incident flux of $5.79 \times 10^{19} \ \mathrm{D \ m^{-2} \ s^{-1}}$ was turned on for $72 \ \mathrm{h}$ with an implantation temperature of $370 \ \mathrm{K}$. The sample then rested for $12 \ \mathrm{h}$ at $295 \ \mathrm{K}$ before beginning the TDS measurement at $300 \ \mathrm{K}$ with a temperature ramp of $0.05 \ \mathrm{K}/s$.
```

To reproduce this experiment, six traps are needed: 1 intrinsic trap and 5 neutron induced traps.

> The trap densities for the neutron induced traps were fitted by {cite}`dark_modelling_2024` for each _dpa_ amount using FESTIM's `NeutronInducedTrap`.

Reviewer: ```suggestion
The trap densities for the neutron induced traps were fitted by {cite}`dark_modelling_2024` for each _dpa_ dose using FESTIM's `NeutronInducedTrap`.
```

> The damage distribution for the neutron-induced traps is as follows:

Reviewer: ```suggestion
The damage distribution for the damage-induced traps is as follows:
```

> $$

Reviewer: ```suggestion

$$
```

    f(x) = \frac{1}{1 + \exp{ \frac{\left( x - x_0 \right)}{\Delta x} }}
$$

Tables with the relevant trap parameters are at the bottom of the page.

+++

## FESTIM code

```{code-cell} ipython3
:tags: [hide-input, hide-output]

# Setup
import festim as F
import numpy as np
import sympy as sp
import matplotlib.pyplot as plt

# Parameters

> # Deuterium in Tungsten [35]

Reviewer: I would avoid hard coding references like this since this index is likely to change. If you want to reference the paper for this diffusivity, please do it in a text block

> D_0 = 1.6 * 1e-7 # m^2 s^-1

Reviewer: ```suggestion
D_0 = 1.6e-7 # m^2 s^-1
```

E_D = 0.28 # eV
w_atom_density = 6.3222e28

> sample_depth = 0.8e-3 # m

Reviewer: I would use `sample_thickness`

> # Table 2

Reviewer: ```suggestion
# Table 2 from Dark et al 10.1088/1741-4326/ad56a0
```

T_imp = 370 # K
T_rest = 295 # K
R_p = 0.7e-9 # m
sigma = 0.5e-9 # m
t_imp = 72 * 3600 # s
t_rest = 12 * 3600 # s
Beta = 0.05 # K s^-1

fluence = 1.5e25
flux = fluence / t_imp

# Deuterium Beam Profile (S = flux * f(x))

> distribution = (1 / (sigma * (2 * np.pi) ** 0.5) * sp.exp(-0.5 * ((F.x - R_p) / sigma) ** 2))

Reviewer: I would maybe define this right before the source

model.materials = damaged_tungsten

# hydrogen source
ion_flux = sp.Piecewise((flux*distribution, F.t < t_imp), (0, True))
source_term = F.Source(value=ion_flux, volume=1, field=0)
model.sources = [source_term]

# boundary conditions
model.boundary_conditions = [F.DirichletBC(surfaces=[1, 2], value=0, field="solute")]

# temperature
start_tds = t_imp + t_rest  # s
min_temp, max_temp = 300, 1000

model.T = F.Temperature(
    value=sp.Piecewise(
        (T_imp, F.t < t_imp),
        (T_rest, F.t < start_tds),
        (min_temp + Beta * (F.t - start_tds), True),
    )
)

# trap settings
k_0 = D_0 / (1.1e-10**2 * 6 * w_atom_density)
damage_dist = 1 / (1 + sp.exp((F.x - 2.5e-06) / 5e-07))
n_i = np.array([2.4e-3, 4.8, 3.8, 2.6, 3.6, 1.1]) * 1e25
neutron_induced_traps = []
E_p = [1.04, 1.15, 1.35, 1.65, 1.85, 2.05]

trap_1 = F.Trap(
    E_k = 0.28,
    k_0 = k_0,
    E_p = 1.04,
    p_0 = 1e13,
    density = n_i[0],
    materials = damaged_tungsten
)

for i in range(1, 6):
    neutron_induced_traps.append(F.Trap(
        k_0 = k_0,
        E_k = E_D,
        p_0 = 1e13,
        E_p = E_p[i],
        density = n_i[i] * damage_dist,
        materials = damaged_tungsten,
    ))

> traps = [trap_1] + neutron_induced_traps
> model.traps = traps

Reviewer: Can you condense this to

```suggestion
model.traps = [trap_1] + neutron_induced_traps
```

model.dt = F.Stepsize(
    initial_value=1,
    stepsize_change_ratio=1.1,
    t_stop=t_imp + t_rest * 0.5,
    dt_min=1e-1,

>     stepsize_stop_max=50,

Reviewer: This triggers a deprecation warning:

```
/home/docs/checkouts/readthedocs.org/user_builds/festim-vv-report/conda/42/lib/python3.11/site-packages/festim/stepsize.py:54: DeprecationWarning: stepsize_stop_max and t_stop attributes will be deprecated in a future release, please use max_stepsize instead
  warnings.warn(
```

Instead please use the new `max_stepsize` parameter (see [here](https://festim.readthedocs.io/en/stable/userguide/stepsize.html#stepsize) for more details)

)

model.settings = F.Settings(
    absolute_tolerance=1e10,
    relative_tolerance=1e-10,
    final_time=start_tds + (max_temp - min_temp) / Beta,  # time to reach max temp
)

derived_quantities = F.DerivedQuantities(
    [
        F.AverageVolume("T", volume=1),
        F.TotalVolume("solute", volume=1),
        F.TotalVolume("retention", volume=1),
        F.TotalVolume("1", volume=1),
        F.TotalVolume("2", volume=1),
        F.TotalVolume("3", volume=1),
        F.TotalVolume("4", volume=1),
        F.TotalVolume("5", volume=1),
        F.TotalVolume("6", volume=1),
        F.HydrogenFlux(surface=1),
        F.HydrogenFlux(surface=2),
    ],
)

model.exports = [derived_quantities]

model.initialise()
model.run()
```

```{note}

> Implementation details from {cite}`dark_modelling_2024_code`.

Reviewer: In reality the implementation details are available in the paper, this will be a copy of the initial implementation. Since you will add the other doses, I would rephrase this as "Experimental data taken from [...]"

```

+++

## Comparison with experimental data

The results produced by FESTIM are in good agreement with the experimental data. The grey areas represent the contribution of each trap to the global TDS spectrum.

```{code-cell} ipython3
:tags: [hide-input]

t = derived_quantities.t
flux_left = derived_quantities.filter(fields="solute", surfaces=1).data
flux_right = derived_quantities.filter(fields="solute", surfaces=2).data
flux_total = -np.array(flux_left) - np.array(flux_right)

t = np.array(t)
temp = min_temp + Beta * (t - start_tds)

# plotting simulation data
plt.plot(temp, flux_total, linewidth=3, label="FESTIM")

# plotting trap contributions
trap_data = [derived_quantities.filter(fields=f"{i}").data for i in range(1, 7)]
contributions = [-np.diff(trap) / np.diff(t) for trap in trap_data]

> colors = [(0.9*(i % 2), 0.2*(i % 4), 0.4*(i % 3)) for i in range(6)]

Reviewer: Just sharing an awesome tool made by @holtzy: [PyPalettes](https://python-graph-gallery.com/color-palette-finder/) which helps finding cool colour palettes!

I quite like [this one](https://python-graph-gallery.com/color-palette-finder/?palette=acadia) for instance.

The python code for this is:

```python
from pypalettes import load_cmap
cmap = load_cmap("Acadia")
```

Author: That is very cool, thanks!

for i, cont in enumerate(contributions):
    if(i == 0):
        label = "Trap 1"
    else:
        label = f"Trap D{i}"

    plt.plot(temp[1:], cont, linestyle="--", color=colors[i], label=label)
    plt.fill_between(temp[1:], 0, cont, facecolor="grey", alpha=0.1)

# plotting original data
experimental_tds = np.genfromtxt("dark-original.csv", delimiter=",")
experimental_temp = experimental_tds[:, 0]
experimental_flux = experimental_tds[:, 1]
plt.scatter(experimental_temp, experimental_flux, color="black", label="original", s=16)

plt.legend()
plt.xlim(min_temp, max_temp)
plt.ylim(bottom=0, top=1e17)
plt.ylabel(r"Desorption flux (m$^{-2}$ s$^{-1}$)")
plt.xlabel(r"Temperature (K)")

plt.show()
```

```{note}
The experimental data was taken from {cite}`dark_modelling_2024` using [WebPlotDigitizer](https://automeris.io/).

> ```

Reviewer: You can use the actual data from [this folder](https://github.com/J-Dark-PhD/Modelling_neutron_damage_effects_in_tungsten/tree/v1.0/section_3_model_parameter_evaluation/data/tds_data_schwartz_selinger)

```{code-cell} ipython3
:tags: [hide-cell]

from myst_nb import glue

for i, trap in enumerate(model.traps):
    for key, value in trap.__dict__.items():

>         glue(f'trap{i}{key}', value, display=False)

Reviewer: I would try to format the strings here.

    glue(f'ni{i}', n_i[i], display=False)
```

> The density distribution of the neutron-induced traps is $n_i f(x)$.

Reviewer: ```suggestion
The density distribution of the neutron-induced traps is $n_i \ f(x)$.
```

> |Trap|k_0|E_k|E_p|p_0|n_i|

Reviewer: Units are needed.
Also, does this table support mathtex? Eg. $k_0$

Author: It does. I added the units using it as well.

    if dpa != 0:
        for i in range(1, 6):
            neutron_induced_traps.append(F.Trap(
                k_0 = k_0,
                E_k = E_D,
                p_0 = 1e13,
                E_p = E_p[i],
                density = dpa_n_i[dpa][i - 1] * damage_dist,
                materials = damaged_tungsten,
            ))

    ## can remove once data empty fix is live
    model.traps = [trap_1] + neutron_induced_traps
    derived_quantities = F.DerivedQuantities(
        [
            F.TotalVolume("solute", volume=1),
            F.TotalVolume("retention", volume=1),
            F.HydrogenFlux(surface=1),
            F.HydrogenFlux(surface=2),
        ],
    )
    ##

    model.exports = [derived_quantities]

    model.initialise()
    model.run()

    results[dpa] = {
        "t" : np.array(derived_quantities.t),
        "flux_left" : np.array(derived_quantities.filter(fields="solute", surfaces=1).data),
        "flux_right" : np.array(derived_quantities.filter(fields="solute", surfaces=2).data),
    }
```

```{code-cell} ipython3
:tags: [hide-input]

# Color Bar
from matplotlib import cm, colors

#norm = colors.LogNorm(vmin=min(dpa_values[1:]), vmax=max(dpa_values))

> norm = colors.LogNorm(1, 2)

Reviewer: This makes the curve all look blackish. I would reverse the norm

|D4|{glue:text}`trap4k_0:.2e`|{glue:text}`trap4E_k:.2e`|{glue:text}`trap4E_p:.2e`|{glue:text}`trap4p_0:.2e`|{glue:text}`ni4:.2e`|
|D5|{glue:text}`trap5k_0:.2e`|{glue:text}`trap5E_k:.2e`|{glue:text}`trap5E_p:.2e`|{glue:text}`trap5p_0:.2e`|{glue:text}`ni5:.2e`|

### Density per induced trap for each dpa dose

|Trap|$0.001$|$0.005$|$0.023$|$0.1$|$0.23$|$0.5$|$2.5$|
|:---|:------|:------|:------|:----|:-----|:----|:----|
|D1|{glue:text}`ni_0_0:.2e`|{glue:text}`ni_0_1:.2e`|{glue:text}`ni_0_2:.2e`|{glue:text}`ni_0_3:.2e`|{glue:text}`ni_0_4:.2e`|{glue:text}`ni_0_5:.2e`|{glue:text}`ni_0_6:.2e`|
|D2|{glue:text}`ni_1_0:.2e`|{glue:text}`ni_1_1:.2e`|{glue:text}`ni_1_2:.2e`|{glue:text}`ni_1_3:.2e`|{glue:text}`ni_1_4:.2e`|{glue:text}`ni_1_5:.2e`|{glue:text}`ni_1_6:.2e`|
|D3|{glue:text}`ni_2_0:.2e`|{glue:text}`ni_2_1:.2e`|{glue:text}`ni_2_2:.2e`|{glue:text}`ni_2_3:.2e`|{glue:text}`ni_2_4:.2e`|{glue:text}`ni_2_5:.2e`|{glue:text}`ni_2_6:.2e`|
|D4|{glue:text}`ni_3_0:.2e`|{glue:text}`ni_3_1:.2e`|{glue:text}`ni_3_2:.2e`|{glue:text}`ni_3_3:.2e`|{glue:text}`ni_3_4:.2e`|{glue:text}`ni_3_5:.2e`|{glue:text}`ni_3_6:.2e`|

> |D5|{glue:text}`ni_4_0:.2e`|{glue:text}`ni_4_1:.2e`|{glue:text}`ni_4_2:.2e`|{glue:text}`ni_4_3:.2e`|{glue:text}`ni_4_4:.2e`|{glue:text}`ni_4_5:.2e`|{glue:text}`ni_4_6:.2e`|

Reviewer: We need to modify this as it's a bit redundant.

Just two tables:

1) for each damage induced trap, all the trap parameters except for density
2) for each damage level, for each trap, density

Author: Do you want me to re-add the tables with this configuration or leave them out? I could also do the data frames experiment now.

Reviewer: As you prefer :-)

Author: I'm trying out the pandas tables, sadly I can't figure out how to display some of their text as latex. They can be converted to markdown, though, so that's a way if we really need it.

Reviewer: This seems related to https://github.com/executablebooks/jupyter-book/issues/1501
